### PR TITLE
Use `AllowSharedBufferSource` for `MLGraphBuilder.constant()`

### LINKED ISSUE #788: `MLGraphBuilder.constant()` should take an `AllowSharedBufferSource`
Currently, `constant()` takes an `ArrayBufferView` (https://www.w3.org/TR/webnn/#ref-for-dom-mlgraphbuilder-constant)

`AllowSharedBufferSource` is specified [here](https://webidl.spec.whatwg.org/#AllowSharedBufferSource) as the following:
```
typedef (ArrayBuffer or SharedArrayBuffer or [AllowShared] ArrayBufferView) AllowSharedBufferSource;
```

Since this buffer data is always immediately copied (https://www.w3.org/TR/webnn/#dom-mlgraphbuilder-constant) I see no reason why we shouldn't allow passing an `ArrayBuffer`, `SharedArrayBuffer` or shared `ArrayBufferView` to this method, as well.

### PR DESCRIPTION
**Pull request by huningxin**
Fix #788

***
<a href="https://pr-preview.s3.amazonaws.com/huningxin/webnn/pull/790.html" title="Last updated on Nov 27, 2024, 6:21 AM UTC (0bd2d0a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/790/c237cf1...huningxin:0bd2d0a.html" title="Last updated on Nov 27, 2024, 6:21 AM UTC (0bd2d0a)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -2012,8 +2013,7 @@ partial dictionary MLOpSupportLimits {
             input, builder.constant(input.dataType, options.minValue));
         } else {
           return builder.min(
-            builder.max(
-              input, builder.constant(input.dataType, options.minValue)),
+            builder.max(input, builder.constant(input.dataType, options.minValue)),
```
Comment (huningxin) [inline index.bs:2016]: This change and following ones seem to be made by tools/reformat-js.py.

File: index.bs [hunk 2]
```diff
@@ -1341,7 +1341,8 @@ interface MLGraphBuilder {
   MLOperand input(USVString name, MLOperandDescriptor descriptor);
 
   // Create an operand for a graph constant.
-  MLOperand constant(MLOperandDescriptor descriptor, ArrayBufferView bufferView);
+  MLOperand constant(MLOperandDescriptor descriptor,
+                     [AllowShared] ArrayBufferView bufferView);
```
Comment (a-sully) [inline index.bs:1345]: Is there a reason we shouldn't use [`AllowSharedBufferSource`](https://webidl.spec.whatwg.org/#AllowSharedBufferSource)?

File: index.bs (hunk 1 above)
Comment (a-sully) [reply index.bs:2016]: Ack. Seems fine to me

File: index.bs (hunk 2 above)
Comment (a-sully) [reply index.bs:1345]: SGTM. I've updated the algorithm in #787 to take an `AllowSharedBufferSource`. Feel free to update this PR accordingly (I don't think this PR actually has to block on #787, but up to you)

File: index.bs [hunk 3]
```diff
@@ -1503,9 +1504,6 @@ Build a composed graph up to a given output operand into a computational graph a
     1. Set |graph|.{{MLGraph/[[context]]}} to [=this=].{{MLGraphBuilder/[[context]]}}.
     1. [=set/For each=] |operand| in |inputs|:
         1. Set |graph|.{{MLGraph/[[inputDescriptors]]}}[|operand|.{{MLOperand/[[name]]}}] to |operand|.{{MLOperand/[[descriptor]]}}.
-
-            Issue(566): If {{MLGraphBuilder/constant(descriptor, bufferView)|constants'}} {{ArrayBuffer}}s are not [=ArrayBuffer/transferred=], make copies for [=MLGraphBuilder/graph=]'s [=computational graph/constants=] here.
```
Comment (huningxin) [inline index.bs:1507]: Note: this issue is closed and not relevant.

File: index.bs [hunk 4]
```diff
@@ -1416,26 +1417,26 @@ Create a named {{MLOperand}} based on a descriptor, that can be used as an input
...
 
 <details open algorithm>
   <summary>
-    The <dfn method for=MLGraphBuilder>constant(|descriptor|, |bufferView|)</dfn> method steps are:
+    The <dfn method for=MLGraphBuilder>constant(|descriptor|, |bufferSource|)</dfn> method steps are:
   </summary>
     1. If [=this=].{{MLGraphBuilder/[[hasBuilt]]}} is true, then [=exception/throw=] an "{{InvalidStateError}}" {{DOMException}}.
     1. If [=MLOperandDescriptor/checking dimensions=] given |descriptor| returns false, then [=exception/throw=] a {{TypeError}}.
-    1. If [=validating buffer with descriptor=] given |bufferView| and |descriptor| returns false, then [=exception/throw=] a {{TypeError}}.
+    1. If [=validating buffer with descriptor=] given |bufferSource| and |descriptor| returns false, then [=exception/throw=] a {{TypeError}}.
```
Comment (huningxin) [inline index.bs:1436]: Note: This depends on #787 for supporting `AllowSharedBufferSource` in "validate buffer with descriptor" algorithm.

File: index.bs (hunk 2 above)
Comment (huningxin) [reply index.bs]: Use `AllowSharedBufferSource` in the latest commit. Open a separate comment for #787 dependency. Resolving this one.

File: index.bs [hunk 5]
```diff
@@ -1341,7 +1341,8 @@ interface MLGraphBuilder {
   MLOperand input(USVString name, MLOperandDescriptor descriptor);
 
   // Create an operand for a graph constant.
-  MLOperand constant(MLOperandDescriptor descriptor, ArrayBufferView bufferView);
+  MLOperand constant(MLOperandDescriptor descriptor,
+                     AllowSharedBufferSource bufferSource);
```
Comment (a-sully) [inline index.bs:1345]: What do you think about just using `buffer`, at least in the IDL?
```suggestion
                     AllowSharedBufferSource buffer);
```

**Review by a-sully — APPROVED** [review]
LGTM!

**Comment by huningxin** [comment]
@fdwr , PTAL? 

/cc @inexorabletash

**Review by inexorabletash — COMMENTED** [review]
Overall LGTM, but +1 to a naming suggestion.

After this and #787 go in, probably when compute() is removed, we should review any lingering `ArrayBufferView` references in the spec and update them as appropriate (to be `AllowSharedBufferSource`, replace with `MLTensor`, reword entirely, etc)

File: index.bs (hunk 5 above)
Comment (a-sully) [reply index.bs:1345]: FWIW `MLNamedArrayBufferViews` are slated to be removed alongside `compute()` #791

File: index.bs (hunk 4 above)
Comment (fdwr) [reply index.bs:1436]: https://github.com/webmachinelearning/webnn/pull/787 is now complete.

**Review by fdwr — APPROVED** [review]
👍

File: index.bs (hunk 5 above)
Comment (huningxin) [reply index.bs]: Thanks for the suggestion and discussion! The latest commit uses `buffer`.

**Review by fdwr — APPROVED** [review]

**Review by a-sully — APPROVED** [review]
Still LGTM 👍

**Review by inexorabletash — APPROVED** [review]
LGTM!

**Comment by huningxin** [comment]
Thanks for the review and approval. I am going to merge it.

**Pull request merged**